packages/core/src/external/carequality/ihe-gateway-v2/monitor/store.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  OutboundPatientDiscoveryReq,
+  OutboundDocumentQueryReq,
+  OutboundDocumentRetrievalReq,
+  XCPDGateway,
+  XCAGateway,
+} from "@metriport/ihe-gateway-sdk";
+import { S3Utils } from "../../../aws/s3";
+import { Config } from "../../../../util/config";
+
+const bucket = Config.getIheResponsesBucketName();
+let s3UtilsInstance = new S3Utils(Config.getAWSRegion());
+
+function getS3UtilsInstance(): S3Utils {
+  return s3UtilsInstance;
+}
+export function setS3UtilsInstanceForStoringIheResponse(s3Utils: S3Utils): void {
+  s3UtilsInstance = s3Utils;
+}
+
+export async function storeXcpdResponses({
+  response,
+  outboundRequest,
+  gateway,
+}: {
+  response: string;
+  outboundRequest: OutboundPatientDiscoveryReq;
+  gateway: XCPDGateway;
+}) {
+  if (!bucket) {
+    return;
+  }
+  const s3Utils = getS3UtilsInstance();
+  const { cxId, patientId, id: requestId } = outboundRequest;
+  const key = buildIheResponseKey({ type: "xcpd", cxId, patientId, requestId, oid: gateway.oid });
+  await s3Utils.uploadFile({
+    bucket,
+    key,
+    file: Buffer.from(response),
+    contentType: "application/xml",
+  });
+}
+
+export async function storeDqResponses({
+  response,
+  outboundRequest,
+  gateway,
+}: {
+  response: string;
+  outboundRequest: OutboundDocumentQueryReq;
+  gateway: XCAGateway;
+}) {
+  if (!bucket) {
+    return;
+  }
+  const s3Utils = getS3UtilsInstance();
+  const { cxId, patientId, id: requestId } = outboundRequest;
+  const key = buildIheResponseKey({
+    type: "dq",
+    cxId,
+    patientId,
+    requestId,
+    oid: gateway.homeCommunityId,
+  });
+  await s3Utils.uploadFile({
+    bucket,
+    key,
+    file: Buffer.from(response),
+    contentType: "application/xml",
+  });
+}
+
+export async function storeDrResponses({
+  response,
+  outboundRequest,
+  gateway,
+}: {
+  response: Buffer;
+  outboundRequest: OutboundDocumentRetrievalReq;
+  gateway: XCAGateway;
+}) {
+  if (!bucket) {
+    return;
+  }
+  const s3Utils = getS3UtilsInstance();
+  const { cxId, patientId, id: requestId } = outboundRequest;
+  const key = buildIheResponseKey({
+    type: "dr",
+    cxId,
+    patientId,
+    requestId,
+    oid: gateway.homeCommunityId,
+  });
+  await s3Utils.uploadFile({ bucket, key, file: response, contentType: "application/xml" });
+}
+
+function buildIheResponseKey({
+  type,
+  cxId,
+  patientId,
+  requestId,
+  oid,
+}: {
+  type: "xcpd" | "dq" | "dr";
+  cxId: string;
+  patientId: string;
+  requestId: string;
+  oid: string;
+}) {
+  return `${type}/${cxId}/${patientId}/${requestId}/${oid}.xml`;
```

Comment:
Also, I would organize the data with Ops in mind, so one is able to get all files for a single requestId easily. And start with the cxId, like to do for regular documents.

Can we have the date as prefix of the file name? That's the only way one can filter files per date on S3, and we prob will want to keep usage there to a limit (have a sweeper to remove files older than a 1 year, for example).

Sugested format:

```ts
`${cxId}/${patientId}/${date}_${requestId}/${type}/${oid}.xml`
```

Reply:
The reqId for xcpd is not the same as the one for dq and dr. How would that change what you are suggesting. 

Date prefix sounds good

Reply:
Oh, good point!

We discussed on Slack and decided on

```ts
${cxId}/${patientId}/{type}/${requestId_date}/${oid}.xml
```